### PR TITLE
soundwire: stream: fix a port param configturation error in multi-pip…

### PR DESCRIPTION
**Pull request by RanderWang**
…elines test

In multi-pipelines test case, there are two pipelines for playback and
capture on the same link bus. The test is failed with a error when
setting port params. The kernel log is included.

[  599.224812] rt711 sdw:0:25d:711:0: invalid dpn_prop direction 1 port_num 0
[  599.224815] sdw_program_slave_port_params failed -22
[  599.224819] intel-sdw sdw-master-0: Program transport params failed: -22
[  599.224822] intel-sdw sdw-master-0: Program params failed: -22
[  599.224828] sdw_enable_stream: SDW0 Pin2-Playback: done

It is a multi-thread issue, the calling sequence is:
(1) hw_params for playback. The playback stream provide the port
    information to Bus.
(2) stream_prepare for playback, Transport and port parameters
    are computed for playback.
(3) hw_params for capture. The capture stream provide the port
    information to Bus.
(4) stream_enable for playback. Program transport and port parameters
    for all masters and slaves. since the transport and port parameters
    are not computed for capture stream, sdw_program_slave_port_params
    will generate a error when setting port params for capture.

This patch ignore the unprepared ports so that the ready ports can work

Tested on comet lake

Signed-off-by: Rander Wang <rander.wang@intel.com>

### CODE REVIEW

File: drivers/soundwire/stream.c [hunk 1]
```diff
@@ -136,8 +136,12 @@ static int sdw_program_slave_port_params(struct sdw_bus *bus,
 	dpn_prop = sdw_get_slave_dpn_prop(s_rt->slave,
 					  s_rt->direction,
 					  t_params->port_num);
-	if (!dpn_prop)
-		return -EINVAL;
+	if (!dpn_prop) {
```
Comment (bardliao) [inline drivers/soundwire/stream.c]: @RanderWang Could you add a comment to describe why dpn_prop is null is not an error here?

Comment (plbossart) [reply drivers/soundwire/stream.c]: Good analysis @RanderWang but we should maybe check what the stream status to decide if that port needs to be programmed first?

Removing an error case doesn't seem quite right.

Comment (RanderWang) [reply drivers/soundwire/stream.c]: @plbossart  it is not feasible to check stream status.  we first program port in _sdw_prepare_stream at https://github.com/thesofproject/linux/blob/integration/soundwire-intel/drivers/soundwire/stream.c#L1504,  then stream->state is set to SDW_STREAM_PREPARED at https://github.com/thesofproject/linux/blob/integration/soundwire-intel/drivers/soundwire/stream.c#L1534. So When setting port params, the stream status may be SDW_STREAM_CONFIGURED, not SDW_STREAM_PREPARED.

**Comment by RanderWang** [comment]
@plbossart  @bardliao  update my PR, thanks!

File: drivers/soundwire/stream.c [hunk 2]
```diff
@@ -267,6 +267,13 @@ static int sdw_program_port_params(struct sdw_master_runtime *m_rt)
 	/* Program transport & port parameters for Slave(s) */
 	list_for_each_entry(s_rt, &m_rt->slave_rt_list, m_rt_node) {
 		list_for_each_entry(p_rt, &s_rt->port_list, port_node) {
+			/*
+			 * skip unprepared slaves since transport & port
+			 * parameters are not computed for them
+			 */
+			if (!p_rt->num)
+				continue;
+
```
Comment (plbossart) [inline drivers/soundwire/stream.c:276]: @RanderWang I don't think this is the right fix, it looks instead as if the data structures are incorrect.

We should have a master_runtime per stream, and then the list of slaves that are used, and last the ports. So when we program the port parameters, we should be able to filter out the entire master runtime.

Also in some cases we will use Port0 for transmission of audio data in BRA mode, so we should not test for port_num != 0.

What you've done is very good, finding the root cause was a very significant progress, but this fix can be improved further.

Comment (RanderWang) [reply drivers/soundwire/stream.c:276]: @plbossart    please check the following description https://github.com/thesofproject/linux/blob/integration/soundwire-intel/drivers/soundwire/bus.h#L81
````
* @num: Port number. For audio streams, valid port number ranges from [1,14]
````
I need your help to make it clear.  Thanks!

Comment (RanderWang) [reply drivers/soundwire/stream.c:276]: I want to get a elegant method to address this issue. First, I added a flag to indicate whether the port is initialized or not. After discussing it with bard,  I abandoned it and also many other changes. I will update my PR. Thanks for review.

**Comment by RanderWang** [comment]
@plbossart   I refined my PR. thanks!

**Comment by plbossart** [comment]
@RanderWang I am not very hot for this suggested fix. The problem is that we use all runtimes at the bus level, without checking the state. Can you try the following diff?

````diff
diff --git a/drivers/soundwire/stream.c b/drivers/soundwire/stream.c
index e95464e46d41..e660c7325fe6 100644
--- a/drivers/soundwire/stream.c
+++ b/drivers/soundwire/stream.c
@@ -634,12 +634,17 @@ static int sdw_notify_config(struct sdw_master_runtime *m_rt)
  *
  * @bus: SDW bus instance
  */
-static int sdw_program_params(struct sdw_bus *bus)
+static int sdw_program_params(struct sdw_bus *bus, bool prepare)
 {
        struct sdw_master_runtime *m_rt;
        int ret = 0;
 
        list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) {
+
+               if (!prepare &&
+                   m_rt->stream->state == SDW_STREAM_CONFIGURED)
+                       continue;
+
                ret = sdw_program_port_params(m_rt);
                if (ret < 0) {
                        dev_err(bus->dev,
@@ -1532,7 +1537,7 @@ static int _sdw_prepare_stream(struct sdw_stream_runtime *stream,
 
 program_params:
                /* Program params */
-               ret = sdw_program_params(bus);
+               ret = sdw_program_params(bus, true);
                if (ret < 0) {
                        dev_err(bus->dev, "Program params failed: %d\n", ret);
                        goto restore_params;
@@ -1635,7 +1640,7 @@ static int _sdw_enable_stream(struct sdw_stream_runtime *stream)
                bus = m_rt->bus;
 
                /* Program params */
-               ret = sdw_program_params(bus);
+               ret = sdw_program_params(bus, false);
                if (ret < 0) {
                        dev_err(bus->dev, "Program params failed: %d\n", ret);
                        return ret;
@@ -1722,7 +1727,7 @@ static int _sdw_disable_stream(struct sdw_stream_runtime *stream)
                struct sdw_bus *bus = m_rt->bus;
 
                /* Program params */
-               ret = sdw_program_params(bus);
+               ret = sdw_program_params(bus, false);
                if (ret < 0) {
                        dev_err(bus->dev, "Program params failed: %d\n", ret);
                        return ret;
@@ -1817,7 +1822,7 @@ static int _sdw_deprepare_stream(struct sdw_stream_runtime *stream)
                }
 
                /* Program params */
-               ret = sdw_program_params(bus);
+               ret = sdw_program_params(bus, false);
                if (ret < 0) {
                        dev_err(bus->dev, "Program params failed: %d\n", ret);
                        return ret;
````

[program_params.diff.txt](https://github.com/thesofproject/linux/files/4030994/program_params.diff.txt)

**Comment by RanderWang** [comment]
@plbossart  thanks, this is really a 

> @RanderWang I am not very hot for this suggested fix. The problem is that we use all runtimes at the bus level, without checking the state. Can you try the following diff?
> 
> ```diff
> diff --git a/drivers/soundwire/stream.c b/drivers/soundwire/stream.c
> index e95464e46d41..e660c7325fe6 100644
> --- a/drivers/soundwire/stream.c
> +++ b/drivers/soundwire/stream.c
> @@ -634,12 +634,17 @@ static int sdw_notify_config(struct sdw_master_runtime *m_rt)
>   *
>   * @bus: SDW bus instance
>   */
> -static int sdw_program_params(struct sdw_bus *bus)
> +static int sdw_program_params(struct sdw_bus *bus, bool prepare)
>  {
>         struct sdw_master_runtime *m_rt;
>         int ret = 0;
>  
>         list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) {
> +
> +               if (!prepare &&
> +                   m_rt->stream->state == SDW_STREAM_CONFIGURED)
> +                       continue;
> +
>                 ret = sdw_program_port_params(m_rt);
>                 if (ret < 0) {
>                         dev_err(bus->dev,
> @@ -1532,7 +1537,7 @@ static int _sdw_prepare_stream(struct sdw_stream_runtime *stream,
>  
>  program_params:
>                 /* Program params */
> -               ret = sdw_program_params(bus);
> +               ret = sdw_program_params(bus, true);
>                 if (ret < 0) {
>                         dev_err(bus->dev, "Program params failed: %d\n", ret);
>                         goto restore_params;
> @@ -1635,7 +1640,7 @@ static int _sdw_enable_stream(struct sdw_stream_runtime *stream)
>                 bus = m_rt->bus;
>  
>                 /* Program params */
> -               ret = sdw_program_params(bus);
> +               ret = sdw_program_params(bus, false);
>                 if (ret < 0) {
>                         dev_err(bus->dev, "Program params failed: %d\n", ret);
>                         return ret;
> @@ -1722,7 +1727,7 @@ static int _sdw_disable_stream(struct sdw_stream_runtime *stream)
>                 struct sdw_bus *bus = m_rt->bus;
>  
>                 /* Program params */
> -               ret = sdw_program_params(bus);
> +               ret = sdw_program_params(bus, false);
>                 if (ret < 0) {
>                         dev_err(bus->dev, "Program params failed: %d\n", ret);
>                         return ret;
> @@ -1817,7 +1822,7 @@ static int _sdw_deprepare_stream(struct sdw_stream_runtime *stream)
>                 }
>  
>                 /* Program params */
> -               ret = sdw_program_params(bus);
> +               ret = sdw_program_params(bus, false);
>                 if (ret < 0) {
>                         dev_err(bus->dev, "Program params failed: %d\n", ret);
>                         return ret;
> ```
> 
> [program_params.diff.txt](https://github.com/thesofproject/linux/files/4030994/program_params.diff.txt)

Thanks!  this is really a magic fix. I misunderstood the master runtime and assumed one bus with one master runtime

**Comment by RanderWang** [comment]
it was included in other PR.

**Comment by plbossart** [comment]
> Thanks! this is really a magic fix. I misunderstood the master runtime and assumed one bus with one master runtime

Yes, it's difficult to detect that the code loops on all the runtimes for each bus, that is very hard to find indeed.

We need to double-check this fix is indeed correct, and we should also check if it's really necessary to program those ports all the time, for prepare and enable.

**Pull request closed without merging**